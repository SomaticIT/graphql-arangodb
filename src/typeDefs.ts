export const directiveTypeDefs = `
"""Represents the direction of an edge in the graph relative to the current node"""
enum AqlEdgeDirection {
  OUTBOUND
  INBOUND
  ANY
}

"""Represents the order of a sorting operation"""
enum AqlSortOrder {
  DESC
  ASC
}

input AqlSortInput {
  """The property to sort on"""
  property: String!
  """The order to sort in. Defaults ASC"""
  order: AqlSortOrder = ASC
  """Change the object being sorted. Defaults to $field"""
  sortOn: String
}

input AqlLimitInput {
  """The upper limit of documents to return"""
  count: String!
  """The number of documents to skip"""
  skip: String
}

directive @document(
  collection: String!
  key: String
  filter: String
  sort: AqlSortInput
  limit: AqlLimitInput
) on FIELD_DEFINITION

directive @node(
  edgeCollection: String!
  direction: AqlEdgeDirection!
  filter: String
  sort: AqlSortInput
  limit: AqlLimitInput
) on FIELD_DEFINITION

directive @edge(
  direction: AqlEdgeDirection!
  collection: String!
  filter: String
  sort: AqlSortInput
  limit: AqlLimitInput
) on FIELD_DEFINITION

directive @edgeNode on FIELD_DEFINITION

directive @aql(
  expression: String!
) on FIELD_DEFINITION

directive @subquery(
  query: String!
) on FIELD_DEFINITION

directive @key on FIELD_DEFINITION

directive @relayConnection(
  edgeCollection: String!
  edgeDirection: AqlEdgeDirection!
  cursorProperty: String!
) on FIELD_DEFINITION | OBJECT

directive @relayEdges on FIELD_DEFINITION | OBJECT

directive @relayPageInfo on FIELD_DEFINITION | OBJECT

directive @relayNode on FIELD_DEFINITION | OBJECT
`;
