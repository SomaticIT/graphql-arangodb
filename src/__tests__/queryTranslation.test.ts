import { makeExecutableSchema } from 'graphql-tools';
import typeDefs from './fixtures/typeDefs';
import { Database } from 'arangojs';
import { graphql } from 'graphql';
import aqlResolver from '..';

describe('query translation integration tests', () => {
  const schema = makeExecutableSchema({
    typeDefs,
    resolvers: {
      Query: {
        user: aqlResolver,
        users: aqlResolver,
        authorizedPosts: aqlResolver,
      },
      Mutation: {
        createUser: async (parent: any, args: any, ctx: any, info: any) => {
          const bindVars = {
            userId: 'foobar',
            role: 'captain',
            name: 'Bob',
          };

          return aqlResolver.runCustomQuery({
            queryString: `
              INSERT {_key: @userId, role: @role, name: @name} INTO users
              RETURN NEW
            `,
            bindVars,
            parent,
            context: ctx,
            info,
          });
        },
        createPost: aqlResolver,
      },
      CreatePostPayload: {
        post: aqlResolver,
      },
    },
  });

  const mockRunQuery = jest.fn();

  const mockDb = ({
    query: mockRunQuery,
  } as any) as Database;

  const run = async (
    query: string,
    mockResults: any[],
    contextValue: any = {}
  ) => {
    mockResults.forEach(mockResult => {
      mockRunQuery.mockResolvedValueOnce({
        all: () => Promise.resolve([mockResult]),
      });
    });

    const result = await graphql({
      schema,
      source: query,
      contextValue: {
        arangoContext: contextValue,
        arangoDb: mockDb,
      },
    });

    if (result.errors) {
      throw result.errors[0];
    }

    return result.data;
  };

  test('translates a basic document query', async () => {
    await run(
      `
      query GetUser {
        user(id: "foo") {
          id
          name
          bio
        }
      }
    `,
      [
        {
          id: 'foo',
          name: 'Foo',
          bio: 'No thanks',
        },
      ]
    );

    expect(mockRunQuery.mock.calls[0][0].query).toMatchSnapshot();
    expect(mockRunQuery.mock.calls[0][0].bindVars).toMatchSnapshot();
  });

  test('translates a document with a nested node', async () => {
    await run(
      `
      query GetUserAndPosts {
        user(id: "foo") {
          id
          name

          simplePosts {
            id
            title
          }
        }
      }
      `,
      [
        {
          id: 'foo',
          name: 'Foo',
          simplePosts: [
            {
              id: 'a',
              title: 'Hello world',
            },
            {
              id: 'b',
              title: 'Hello again world',
            },
            {
              id: 'c',
              title: 'Come here often, world?',
            },
          ],
        },
      ]
    );

    expect(mockRunQuery.mock.calls[0][0].query).toMatchSnapshot();
    expect(mockRunQuery.mock.calls[0][0].bindVars).toMatchSnapshot();
  });

  test('filters', async () => {
    await run(
      `
      query GetUserAndFilteredPosts {
        user(id: "foo") {
          id

          filteredPosts(titleMatch: "here") {
            id
            title
          }
        }
      }
      `,
      [
        {
          id: 'foo',
          filteredPosts: [
            {
              id: 'c',
              title: 'Come here often, world?',
            },
          ],
        },
      ]
    );

    expect(mockRunQuery.mock.calls[0][0].query).toMatchSnapshot();
    expect(mockRunQuery.mock.calls[0][0].bindVars).toMatchSnapshot();
  });

  test('paginates', async () => {
    await run(
      `
      query GetUserAndPaginatedPosts {
        user(id: "foo") {
          id

          paginatedPosts(count: 2) {
            id
            title
          }
        }
      }
      `,
      [
        {
          id: 'foo',
          paginatedPosts: [
            {
              id: 'b',
              title: 'Hello again world',
            },
            {
              id: 'c',
              title: 'Come here often, world?',
            },
          ],
        },
      ]
    );

    expect(mockRunQuery.mock.calls[0][0].query).toMatchSnapshot();
    expect(mockRunQuery.mock.calls[0][0].bindVars).toMatchSnapshot();
  });

  test('sorts descending', async () => {
    await run(
      `
      query GetUserAndDescendingPosts {
        user(id: "foo") {
          id

          descendingPosts {
            id
            title
          }
        }
      }
      `,
      [
        {
          id: 'foo',
          descendingPosts: [
            {
              id: 'b',
              title: 'Hello again world',
            },
            {
              id: 'c',
              title: 'Come here often, world?',
            },
          ],
        },
      ]
    );

    expect(mockRunQuery.mock.calls[0][0].query).toMatchSnapshot();
    expect(mockRunQuery.mock.calls[0][0].bindVars).toMatchSnapshot();
  });

  test('traverses edges', async () => {
    await run(
      `
      query GetUserAndFriends {
        user(id: "foo") {
          id
          name

          friends {
            strength

            user {
              id
              name
            }
          }
        }
      }
      `,
      [
        {
          id: 'foo',
          name: 'Bar',
          friends: [
            {
              strength: 2,
              user: {
                id: 'bar',
                name: 'Jeff',
              },
            },
          ],
        },
      ]
    );

    expect(mockRunQuery.mock.calls[0][0].query).toMatchSnapshot();
    expect(mockRunQuery.mock.calls[0][0].bindVars).toMatchSnapshot();
  });

  test('runs arbitrary subqueries', async () => {
    await run(
      `
      query GetUserAndFriends {
        user(id: "foo") {
          id
          name

          friendsOfFriends {
            id
            name
          }
        }
      }
      `,
      [
        {
          id: 'foo',
          name: 'Bar',
          friendsOfFriends: [
            {
              id: 'baz',
              name: 'Eva',
            },
          ],
        },
      ]
    );

    expect(mockRunQuery.mock.calls[0][0].query).toMatchSnapshot();
    expect(mockRunQuery.mock.calls[0][0].bindVars).toMatchSnapshot();
  });

  test('uses context values', async () => {
    await run(
      `
      query GetAuthorizedPosts {
        authorizedPosts {
          id
          title
        }
      }
      `,
      [
        [
          {
            id: 'a',
            title: 'Hello world',
          },
          {
            id: 'b',
            title: 'Hello again world',
          },
        ],
      ],
      {
        userId: 'foo',
      }
    );

    expect(mockRunQuery.mock.calls[0][0].query).toMatchSnapshot();
    expect(mockRunQuery.mock.calls[0][0].bindVars).toMatchSnapshot();
  });

  test('runs aql expressions', async () => {
    await run(
      `
      query GetUserAndFriends {
        user(id: "foo") {
          id

          fullName
        }
      }
      `,
      [
        {
          id: 'foo',
          fullName: 'Foo Bar',
        },
      ]
    );

    expect(mockRunQuery.mock.calls[0][0].query).toMatchSnapshot();
    expect(mockRunQuery.mock.calls[0][0].bindVars).toMatchSnapshot();
  });

  test('does a Relay-style connection', async () => {
    await run(
      `
      query GetUser {
        user(id: "foo") {
          id

          postsConnection(after: "opaqueCursor", filter: { publishedAfter: "2019-17-08 04:27:54 AM" }) {
            edges {
              cursor
              node {
                id
                title
              }
            }
            pageInfo {
              hasNextPage
            }
          }
        }
      }
    `,
      [
        {
          id: 'foo',

          postsConnection: {
            edges: [
              {
                cursor: 'a',
                node: {
                  id: 'a',
                  title: 'Hello world',
                },
              },
            ],
            pageInfo: {
              hasNextPage: false,
            },
          },
        },
      ]
    );

    expect(mockRunQuery.mock.calls[0][0].query).toMatchSnapshot();
    expect(mockRunQuery.mock.calls[0][0].bindVars).toMatchSnapshot();
  });

  test('passes traversal options', async () => {
    await run(
      `
      query GetUserAndPosts {
        user(id: "foo") {
          id
          name

          bfsPosts {
            id
            title
          }
        }
      }
      `,
      [
        {
          id: 'foo',
          name: 'Foo',
          bfsPosts: [
            {
              id: 'a',
              title: 'Hello world',
            },
            {
              id: 'b',
              title: 'Hello again world',
            },
            {
              id: 'c',
              title: 'Come here often, world?',
            },
          ],
        },
      ]
    );

    expect(mockRunQuery.mock.calls[0][0].query).toMatchSnapshot();
    expect(mockRunQuery.mock.calls[0][0].bindVars).toMatchSnapshot();
  });

  test('resolves a custom query inside the resolver with selections', async () => {
    await run(
      `
      mutation CreateUser {
        createUser {
          id
          name

          simplePosts {
            id
            title
          }
        }
      }
      `,
      [
        {
          id: 'foobar',
          name: 'Bob',
          simplePosts: [],
        },
      ]
    );

    expect(mockRunQuery.mock.calls[0][0].query).toMatchSnapshot();
    expect(mockRunQuery.mock.calls[0][0].bindVars).toMatchSnapshot();
  });

  test('resolves multi-query operations to avoid read-after-write errors', async () => {
    const result = await run(
      `
      mutation CreatePost {
        createPost {
          post {
            id
            title
            author {
              id
            }
          }
        }
      }
      `,
      [
        {
          postId: '3',
        },
        {
          id: '3',
          title: 'Fake post',
          author: {
            id: 'foo',
          },
        },
      ]
    );

    expect(result).toEqual({
      createPost: {
        post: {
          id: '3',
          title: 'Fake post',
          author: {
            id: 'foo',
          },
        },
      },
    });

    expect(mockRunQuery).toHaveBeenCalledTimes(2);
    expect(mockRunQuery.mock.calls[0][0].query).toMatchSnapshot();
    expect(mockRunQuery.mock.calls[0][0].bindVars).toMatchSnapshot();
    expect(mockRunQuery.mock.calls[1][0].query).toMatchSnapshot();
    expect(mockRunQuery.mock.calls[1][0].bindVars).toMatchSnapshot();
  });
});
