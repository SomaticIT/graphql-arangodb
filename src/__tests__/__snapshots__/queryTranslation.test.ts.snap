// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`query translation integration tests does a Relay-style connection 1`] = `
"LET query = FIRST(
  LET user = DOCUMENT(users, @field_user.args.id)
  RETURN {
    id: user._key,
    postsConnection: FIRST(
      LET user_postsConnection_listPlusOne = (
        FOR user_postsConnection_node, user_postsConnection_edge IN OUTBOUND user posted
          OPTIONS {bfs: true}
        FILTER user_postsConnection_node && (!@field_user_postsConnection.args.after || user_postsConnection_node.title > @field_user_postsConnection.args.after) && (
          @field_user_postsConnection.args['filter'] != null && (
            @field_user_postsConnection.args['filter'].publishedAfter == null || user_postsConnection_node.publishedAt > @field_user_postsConnection.args['filter'].publishedAfter
          ) && (
            @field_user_postsConnection.args['filter'].titleLike == null || LIKE(user_postsConnection_node.title, CONCAT(\\"%\\", @field_user_postsConnection.args['filter'].titleLike, \\"%\\"))
          )
        )
        SORT user_postsConnection_node.title
        LIMIT @field_user_postsConnection.args.first + 1
        RETURN MERGE(user_postsConnection_edge, { cursor: user_postsConnection_node.title, node: user_postsConnection_node })
      )
      LET user_postsConnection = {
        edges: SLICE(user_postsConnection_listPlusOne, 0, @field_user_postsConnection.args.first),
        pageInfo: { 
          hasNextPage: LENGTH(user_postsConnection_listPlusOne) == @field_user_postsConnection.args.first + 1
        }
      }
      RETURN {
        edges: (
          FOR user_postsConnection_edges IN user_postsConnection.edges
          RETURN {
            cursor: user_postsConnection_edges.cursor,
            node: FIRST(
              LET user_postsConnection_edges_node = user_postsConnection_edges.node
              RETURN {
                title: user_postsConnection_edges_node.title,
                id: user_postsConnection_edges_node._key
              }
            )
          }
        ),
        pageInfo: FIRST(
          LET user_postsConnection_pageInfo = user_postsConnection.pageInfo
          RETURN {
            hasNextPage: user_postsConnection_pageInfo.hasNextPage
          }
        )
      }
    )
  }
)
RETURN query"
`;

exports[`query translation integration tests does a Relay-style connection 2`] = `
Object {
  "field_user": Object {
    "args": Object {
      "id": "foo",
    },
  },
  "field_user_postsConnection": Object {
    "args": Object {
      "after": "opaqueCursor",
      "filter": Object {
        "publishedAfter": "2019-17-08 04:27:54 AM",
      },
      "first": 10,
    },
  },
}
`;

exports[`query translation integration tests filters 1`] = `
"LET query = FIRST(
  LET user = DOCUMENT(users, @field_user.args.id)
  RETURN {
    id: user._key,
    filteredPosts: (
      FOR user_filteredPosts IN OUTBOUND user posted
        FILTER user_filteredPosts.title =~ @field_user_filteredPosts.args.titleMatch
      RETURN {
        title: user_filteredPosts.title,
        id: user_filteredPosts._key
      }
    )
  }
)
RETURN query"
`;

exports[`query translation integration tests filters 2`] = `
Object {
  "field_user": Object {
    "args": Object {
      "id": "foo",
    },
  },
  "field_user_filteredPosts": Object {
    "args": Object {
      "titleMatch": "here",
    },
  },
}
`;

exports[`query translation integration tests paginates 1`] = `
"LET query = FIRST(
  LET user = DOCUMENT(users, @field_user.args.id)
  RETURN {
    id: user._key,
    paginatedPosts: (
      FOR user_paginatedPosts IN OUTBOUND user posted
        SORT user_paginatedPosts[@field_user_paginatedPosts.args.sort] ASC
        LIMIT @field_user_paginatedPosts.args.skip @field_user_paginatedPosts.args.count
      RETURN {
        title: user_paginatedPosts.title,
        id: user_paginatedPosts._key
      }
    )
  }
)
RETURN query"
`;

exports[`query translation integration tests paginates 2`] = `
Object {
  "field_user": Object {
    "args": Object {
      "id": "foo",
    },
  },
  "field_user_paginatedPosts": Object {
    "args": Object {
      "count": 2,
      "skip": 0,
      "sort": "title",
    },
  },
}
`;

exports[`query translation integration tests passes traversal options 1`] = `
"LET query = FIRST(
  LET user = DOCUMENT(users, @field_user.args.id)
  RETURN {
    name: user.name,
    id: user._key,
    bfsPosts: (
      FOR user_bfsPosts IN OUTBOUND user posted
        OPTIONS { bfs: true }
      RETURN {
        title: user_bfsPosts.title,
        id: user_bfsPosts._key
      }
    )
  }
)
RETURN query"
`;

exports[`query translation integration tests passes traversal options 2`] = `
Object {
  "field_user": Object {
    "args": Object {
      "id": "foo",
    },
  },
}
`;

exports[`query translation integration tests resolves a custom query inside the resolver with selections 1`] = `
"LET query = FIRST(
  LET createUser = FIRST(
    
                  INSERT {_key: @userId, role: @role, name: @name} INTO users
                  RETURN NEW
                
  )
  RETURN {
    name: createUser.name,
    id: createUser._key,
    simplePosts: (
      FOR createUser_simplePosts IN OUTBOUND createUser posted
      RETURN {
        title: createUser_simplePosts.title,
        id: createUser_simplePosts._key
      }
    )
  }
)
RETURN query"
`;

exports[`query translation integration tests resolves a custom query inside the resolver with selections 2`] = `
Object {
  "name": "Bob",
  "role": "captain",
  "userId": "foobar",
}
`;

exports[`query translation integration tests resolves multi-query operations to avoid read-after-write errors 1`] = `
"LET query = FIRST(
  INSERT { title: \\"Fake post\\", body: \\"foo\\", publishedAt: \\"2019-05-03\\" }
  INTO posts
  OPTIONS { waitForSync: true }
  LET createPost = {
    post: NEW
  }
  RETURN {
    post: createPost.post
  }
)
RETURN query"
`;

exports[`query translation integration tests resolves multi-query operations to avoid read-after-write errors 2`] = `Object {}`;

exports[`query translation integration tests resolves multi-query operations to avoid read-after-write errors 3`] = `
"LET query = FIRST(
  LET post = DOCUMENT(posts, @parent._key)
  RETURN {
    title: post.title,
    id: post._key,
    author: FIRST(
      FOR post_author IN INBOUND post posted
        LIMIT 1
      RETURN {
        id: post_author._key
      }
    )
  }
)
RETURN query"
`;

exports[`query translation integration tests resolves multi-query operations to avoid read-after-write errors 4`] = `
Object {
  "parent": Object {
    "postId": "3",
  },
}
`;

exports[`query translation integration tests runs aql expressions 1`] = `
"LET query = FIRST(
  LET user = DOCUMENT(users, @field_user.args.id)
  RETURN {
    id: user._key,
    fullName: CONCAT(user.name, \\" \\", user.surname)
  }
)
RETURN query"
`;

exports[`query translation integration tests runs aql expressions 2`] = `
Object {
  "field_user": Object {
    "args": Object {
      "id": "foo",
    },
  },
}
`;

exports[`query translation integration tests runs arbitrary subqueries 1`] = `
"LET query = FIRST(
  LET user = DOCUMENT(users, @field_user.args.id)
  RETURN {
    name: user.name,
    id: user._key,
    friendsOfFriends: (
      FOR user_friendsOfFriends IN 2..2 ANY user friendOf OPTIONS {bfs: true, uniqueVertices: 'path'}
      RETURN {
        name: user_friendsOfFriends.name,
        id: user_friendsOfFriends._key
      }
    )
  }
)
RETURN query"
`;

exports[`query translation integration tests runs arbitrary subqueries 2`] = `
Object {
  "field_user": Object {
    "args": Object {
      "id": "foo",
    },
  },
}
`;

exports[`query translation integration tests sorts descending 1`] = `
"LET query = FIRST(
  LET user = DOCUMENT(users, @field_user.args.id)
  RETURN {
    id: user._key,
    descendingPosts: (
      FOR user_descendingPosts IN OUTBOUND user posted
        SORT user_descendingPosts[\\"title\\"] DESC
      RETURN {
        title: user_descendingPosts.title,
        id: user_descendingPosts._key
      }
    )
  }
)
RETURN query"
`;

exports[`query translation integration tests sorts descending 2`] = `
Object {
  "field_user": Object {
    "args": Object {
      "id": "foo",
    },
  },
}
`;

exports[`query translation integration tests translates a basic document query 1`] = `
"LET query = FIRST(
  LET user = DOCUMENT(users, @field_user.args.id)
  RETURN {
    name: user.name,
    bio: user.bio,
    id: user._key
  }
)
RETURN query"
`;

exports[`query translation integration tests translates a basic document query 2`] = `
Object {
  "field_user": Object {
    "args": Object {
      "id": "foo",
    },
  },
}
`;

exports[`query translation integration tests translates a document with a nested node 1`] = `
"LET query = FIRST(
  LET user = DOCUMENT(users, @field_user.args.id)
  RETURN {
    name: user.name,
    id: user._key,
    simplePosts: (
      FOR user_simplePosts IN OUTBOUND user posted
      RETURN {
        title: user_simplePosts.title,
        id: user_simplePosts._key
      }
    )
  }
)
RETURN query"
`;

exports[`query translation integration tests translates a document with a nested node 2`] = `
Object {
  "field_user": Object {
    "args": Object {
      "id": "foo",
    },
  },
}
`;

exports[`query translation integration tests traverses edges 1`] = `
"LET query = FIRST(
  LET user = DOCUMENT(users, @field_user.args.id)
  RETURN {
    name: user.name,
    id: user._key,
    friends: (
      FOR user_friends_node, user_friends IN ANY user friendOf
      RETURN {
        strength: user_friends.strength,
        user: user_friends_node
      }
    )
  }
)
RETURN query"
`;

exports[`query translation integration tests traverses edges 2`] = `
Object {
  "field_user": Object {
    "args": Object {
      "id": "foo",
    },
  },
}
`;

exports[`query translation integration tests uses context values 1`] = `
"LET query = (
  LET authenticatedUser = DOCUMENT('users', @context.userId)
  LET allAuthorizedPosts = UNION_DISTINCT(
    (FOR post IN posts FILTER post.public == true RETURN post),
    (FOR post IN OUTBOUND authenticatedUser posted RETURN post)
  )
  FOR authorizedPosts IN allAuthorizedPosts
  RETURN {
    title: authorizedPosts.title,
    id: authorizedPosts._key
  }
)
RETURN query"
`;

exports[`query translation integration tests uses context values 2`] = `
Object {
  "context": Object {
    "userId": "foo",
  },
}
`;
